sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Command.scala

```diff
@@ -37,3 +38,35 @@ trait Command extends LogicalPlan {
 trait LeafCommand extends Command with LeafLike[LogicalPlan]
 trait UnaryCommand extends Command with UnaryLike[LogicalPlan]
 trait BinaryCommand extends Command with BinaryLike[LogicalPlan]
+
+/**
+ * A logical node that represents a command whose children are only analyzed, but not optimized.
+ */
+trait AnalysisOnlyCommand extends Command {
+  private var _isAnalyzed: Boolean = false
+
+  def childrenToAnalyze: Seq[LogicalPlan]
+
+  override def children: Seq[LogicalPlan] = if (_isAnalyzed) Nil else childrenToAnalyze
```

Comment:
This is tricky because the `children` can change dynamically. I was expecting to put `isAnalyzed` as a case class parameter, so `markAsAnalyzed` creates a new copy and the plan node is still immutable. We can avoid changing `TreeNode` if we do so.

Reply:
Hmm, I don't think we can use case class at this level (e.g., cannot have abstract member like `childrenToAnalyze`), right?

If we need to make the node immutable, I think the responsibility should be at the concrete command - similar to the first commit I had?

Reply:
My proposal was
```
trait AnalysisOnlyCommand extends Command {
  val isAnalyzed: Boolean
  def childrenToAnalyze: Seq[LogicalPlan]
  def children = if (isAnalyzed) Nil else childrenToAnalyze
  def markAsAnalyzed(): LogicalPlan
}

The concrete command needs to implement `markAsAnalyzed` and copy itself with `isAnalyzed=true`. 
```

Reply:
Ah sorry, I misunderstood the original comment. Thanks for the clarification!